src/modules/robot/Robot.cpp

```diff
@@ -1289,73 +1346,58 @@ bool Robot::append_milestone(const float target[], float rate_mm_s)
-            actuator_pos[i] = transformed_target[i];
-        }
-    }
 
-#if MAX_ROBOT_ACTUATORS > 3
-    sos= 0;
-    // for the extruders just copy the position, and possibly scale it from mm³ to mm
-    for (size_t i = E_AXIS; i < n_motors; i++) {
-        actuator_pos[i]= transformed_target[i];
-        if(actuators[i]->is_extruder() && get_e_scale_fnc) {
-            // NOTE this relies on the fact only one extruder is active at a time
-            // scale for volumetric or flow rate
-            // TODO is this correct? scaling the absolute target? what if the scale changes?
-            // for volumetric it basically converts mm³ to mm, but what about flow rate?
-            actuator_pos[i] *= get_e_scale_fnc();
-        }
-        if(auxilliary_move) {
-            // for E only moves we need to use the scaled E to calculate the distance
+    } else {
+        // for solo E and rotary axis moves we need to use the scaled E or ABC to calculate the distance
+        sos= 0;
+        for (size_t i = E_AXIS; i < n_motors; i++) {
             sos += powf(actuator_pos[i] - actuators[i]->get_last_milestone(), 2);
         }
+        distance= sqrtf(sos); // distance in mm of the e or ABC move
+        if(distance < 0.00001F) return false; // avoid divide by zero further on
     }
-    if(auxilliary_move) {
-        distance= sqrtf(sos); // distance in mm of the e move
-        if(distance < 0.00001F) return false;
-    }
-#endif
 
     // use default acceleration to start with
     float acceleration = default_acceleration;
 
-    float isecs = rate_mm_s / distance;
+    // this is the time it should take to execute this cartesian move, it is approximate
+    // as it does not take into account accel/decel
+    float secs = distance / rate_mm_s;
 
-    // check per-actuator speed limits
+    // check per-actuator speed limits by looking at the minimum time each axis can move its specified amount
+    // this is regardless of whether it is mm/sec or deg/sec for a rotary axis
     for (size_t actuator = 0; actuator < n_motors; actuator++) {
+        // actual distance moved for this actuator
+        // NOTE for a rotary axis this will be degrees turned not distance
         float d = fabsf(actuator_pos[actuator] - actuators[actuator]->get_last_milestone());
-        if(d == 0 || !actuators[actuator]->is_selected()) continue; // no movement for this actuator
-
-        float actuator_rate= d * isecs;
-        if (actuator_rate > actuators[actuator]->get_max_rate()) {
-            rate_mm_s *= (actuators[actuator]->get_max_rate() / actuator_rate);
-            isecs = rate_mm_s / distance;
+        if(d < 0.00001F || !actuators[actuator]->is_selected()) continue; // no perceptible movement for this actuator
+
+        // find approximate min time this axis needs to move its distance
+        float actuator_min_time= d / actuators[actuator]->get_max_rate();
+        if (secs < actuator_min_time ) {
```

Comment:
This whole `secs` (or the original `isecs`) calculation is not used anywhere and burns unnecessary fpu cycles. Like I described yesterday, it's all about a simple **ratio** between the Euclidean toolpath `distance` and the axis speed and acceleration. Because the **ratio** is mostly used in division, I implemented it using a `limit_factor` which is simply the inverse of the ratio. It is still 14 times faster to [mulitply than to divide on the fpu](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0439b/BEHJADED.html).

Reply:
what FPU :)

Reply:
of course it is used, it is what handles different actuator speeds that are lower.